### PR TITLE
Allow to configure prefix for internal Kafka fields

### PR DESCRIPTION
**Pull request by ssheikin**
## Description
When we process Kafka topics, for internal usage we add some bunch of additional columns with some useful data and we fill them with data which is coming from Kafka (RecordSet), for example:
`case PARTITION_OFFSET_FIELD -> longValueProvider(message.offset());`
By default in current implementation this columns have some hardcoded names like:
`_partition_id` `_message_corrupt` e.t.c
So it could be a situation, that Kafka topic itself have some fields with similar name, and in this case during processing we could have a conflict (like two columns have the same name, for example `_key`)
So reason of this PR is to provide ability to tune internal column names with custom prefix like `XX_my_prefix_XX_key`, so this conflict could be more rare then with simple current default prefix `_key`
This PR will not resolve problem, but just postpone it until times when users will have more difficult column name, which will conflict with our internal column.
This change is backward compatible so for people who don't have colliding field names they don't need to change their existing queries to work with newer version.
The drawback is that this can't anticipate a collision ahead of time. It does however allow adminstrators to use a unique prefix unlikely to ever exist - e.g. `__kafka_message_metadata_`.

Thank you, @vlad-lyutenko and @hashhar for the description :)

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(*) Release notes are required, with the following suggested text:
Workaround for encountering "multiple entries with same key" while querying Kafka

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaConnectorTest.java
Comment (hashhar) [inline plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaConnectorTest.java:460]: Are these removed because they are unused?

Comment (hashhar) [reply plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaConnectorTest.java:460]: That'll teach me to review commit by commit before entire diff. 🤦

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaInternalFieldManager.java
Comment (hashhar) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaInternalFieldManager.java:163]: we have this to preserve order of columns, correct?

Earlier code the order being preserved was mostly an impl detail of ImmutableMap that insertion order = iteration order?

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:98]: I've played with this code quite a bit, moving parts here and there, but it occurred that rate `cleanness` / (`amount of changes` * `amount of code`) is low. 
E.g. there is type casting which makes `foreach` not laconic.

I'd happily apply your code suggestion.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaInternalFieldManager.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaInternalFieldManager.java:110]: Don't know yet. My feeling is that this code had long and interesting live, and mostly was changed by small chunks.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaInternalFieldManager.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaInternalFieldManager.java:163]: Don't know about ordering. I guess it is not relevant here.

This code is tricky. 
Once it looked like enum already. And was refactored this way to use dynamic value from `TypeManager`.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java
Comment (ssheikin) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:108]: `(coding) InnerAssignment: Inner assignments should be avoided.`
Don't use switch expression instead of switch statement.

**Comment by vlad-lyutenko** [comment]
I tried to understand what is going in this PR, let me try to summarise it and please correct me if I am wrong or miss something.

When we process Kafka topics, for internal usage we add some bunch of additional columns with some useful data and we fill them with data which is coming from Kafka (RecordSet), for example:

`case PARTITION_OFFSET_FIELD -> longValueProvider(message.offset());`

By default in current implementation this columns have some hardcoded names like:

`_partition_id` `_message_corrupt` e.t.c

So it could be a situation, that Kafka topic itself have some fields with similar name, and in this case during processing we could have a conflict (like to columns have the same name, for example `_key`)

So reason of this PR is to provide ability to tune internal column names with custom prefix like `XX_my_prefix_XX_key`,
so this conflict could be more rare then with simple current default prefix `_key`

If my description is correct and I understand correctly, this PR will not resolve problem, but just postpone it until times when users will have more difficult column name, which will conflict with our internal column.

But maybe I missed smth, please correct me if I ma wrong

**Comment by hashhar** [comment]
@vlad-lyutenko Exactly correct understanding.

> this PR will not resolve problem, but just postpone it until times when users will have more difficult column name, which will conflict with our internal column.

True. This change has some benefits - it's backward compatible so for people who don't have colliding field names they don't need to change their existing queries to work with newer version.

The drawback is that this can't anticipate a collision ahead of time. It does however allow adminstrators to use a unique prefix unlikely to ever exist - e.g. `__kafka_message_metadata_`.

Is there some other solution you can think of? One idea which I was thinking was to expose all internal columns as a `ROW` type with multiple fields instead of separate column per field - that way you only have to worry about one collision.

**Comment by vlad-lyutenko** [comment]
> Is there some other solution you can think of? One idea which I was thinking was to expose all internal columns as a `ROW` type with multiple fields instead of separate column per field - that way you only have to worry about one collision.

I think in this case, we will lose backward compatibility for people who already have some queries,
I more think about solution of providing users with more detailed description what happens and how to resolve problem, using custom prefix.
Looks like current exception is coming somewhere from core/engine (not from connector).
So maybe we can add such check on column collision inside connector, with more detailed description, not to get any open issues/tickets from users in future.

**Comment by hashhar** [comment]
> I more think about solution of providing users with more detailed description what happens and how to resolve problem, using custom prefix.

This I agree with - see https://github.com/trinodb/trino/pull/14224#discussion_r976109333

**Comment by Praveen2112** [comment]
What if we could generate the name of these meta-columns based on the table structure. Like if we have column `_timestamp_n` then we could generate the metacolumn name as `_timestamp_(n+1)` it would be backward compatible and will also avoid the conflict if it happens anytime in the future ?

**Comment by hashhar** [comment]
I think the current solution is the best we can do for now while also not breaking user queries.

An alternative might be to expose these columns using prefixes which are invalid for the target system - in Hive that's `$` but AFAIK in Kafka connector there are no limitations on field names other than for Avro which requires prefix to be `[A-Za-z_]`.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java
Comment (hashhar) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:98]: I just meant something like:

```java
if (domains.isPresent()) {
    for (Map.Entry<ColumnHandle, Domain> entry : domains.get().entrySet()) {
        <existing_code>
    }
}
```

But after reading more I see why the older was safe as well:

We `verify(!constraint.isNone())` which guarantees that `domains != Optional.empty()`.
We also are within an `if (!constraint.isAll())` which implies we're guaranteed to have one or more values in entryset.

Comment (hashhar) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:98]: All this to say - no changes requested.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java
Comment (hashhar) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:108]: bump, please rename

**Review by hashhar — APPROVED** [review]

**Comment by hashhar** [comment]
LGTM % remaining comments. Please let me know when addressed @ssheikin

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:108]: I think it's ok to be as it is, because it relates to verb `filter` in passive voice, an if you filter offset by some range it could be `ranged`

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:98]: I had to open my eyes wider and to take a look on the code broader.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaInternalFieldManager.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaInternalFieldManager.java:110]: Just an example: https://github.com/prestodb/presto/pull/11088

**Comment by ssheikin** [comment]
>Looks like current exception is coming somewhere from core/engine (not from connector).
So maybe we can add such check on column collision inside connector, with more detailed description, not to get any open issues/tickets from users in future.

Current exception comes from connector. 
I've Improved exception message.

**Comment by ssheikin** [comment]
@hashhar @vlad-lyutenko @Praveen2112 All comments are addressed. Please take a look one more time.

**Comment by vlad-lyutenko** [comment]
> > Looks like current exception is coming somewhere from core/engine (not from connector).
> > So maybe we can add such check on column collision inside connector, with more detailed description, not to get any open issues/tickets from users in future.
> 
> Current exception comes from connector. I've Improved exception message.

Ok, thx, now I see it was from `buildOrThrow` of immutable map.

**Comment by vlad-lyutenko** [comment]
I am ok with chnages

**Review by vlad-lyutenko — APPROVED** [review]

**Comment by ssheikin** [comment]
ci / pt (default, suite-6-non-generic, ) (pull_request) 
```
tests               | 2022-09-22 23:00:02 INFO: FAILURE     /    io.trino.tests.product.kafka.TestKafkaPushdownSmokeTest.testCreateTimePushdown (Groups: profile_specific_tests, kafka) took 6.0 seconds
tests               | 2022-09-22 23:00:02 SEVERE: Failure cause:
tests               | java.lang.AssertionError: Not equal rows:
tests               | 0 - expected: 2|
tests               | 0 - actual:   3|
tests               | 	at io.trino.tests.product.kafka.TestKafkaPushdownSmokeTest.testCreateTimePushdown(TestKafkaPushdownSmokeTest.java:190)
tests               | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
tests               | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
tests               | 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
tests               | 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
tests               | 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
tests               | 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
tests               | 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
tests               | 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
tests               | 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
tests               | 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
tests               | 	at java.base/java.lang.Thread.run(Thread.java:833)
```

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaRecordSet.java
Comment (hashhar) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaRecordSet.java:116]: any specific reason to move to list from statically defined array?

File: docs/src/main/sphinx/connector/kafka.rst
Comment (hashhar) [inline docs/src/main/sphinx/connector/kafka.rst:256]: Probably move this before the table with something like:

```rst
Note that the internal column name prefix is configurable by setting
``kafka.internal-field-prefix`` catalog configuration property and it defaults
to ``_``. In case you have configured a different value the internal column
names will appear different from the table below.
```

(Users often skim things and don't notice impact of some config unless  you spell it out clearly).

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaConfig.java
Comment (hashhar) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaConfig.java:186]: nit: Since we call them internal columns in user-facing docs should this config be called `kafka.internal-column-prefix`?

**Review by hashhar — APPROVED** [review]
LGTM % nits + question about config name

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaRecordSet.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaRecordSet.java:116]: No, only
> - do not copy currentRowValues unnecessarily

Comment (hashhar) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaRecordSet.java:116]: I don't know if copying is worse or pointer chasing. 🤷 

This is in RecordSet which is most probably performance sensitive. I'll like a second opinion from @Praveen2112 / @raunaqmorarka.

One way to mitigate my concern would be pre-size the ImmutableList instead of empty list.

Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaRecordSet.java:116]: > pointer chasing

It's not obvious how this effect can be obtained/avoided here.
Looks more like legacy code.

> pre-size the ImmutableList

I don't think it matters here.

Comment (raunaqmorarka) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaRecordSet.java:116]: I didn't follow why we changed this code. The existing code was copying object references to a static array. Now we create a new list for each row and then copy object references to it. The array approach seems cheaper, so to me it looks like we should leave it unchanged.
In general if performance is a concern, and the Kafka consumer fetches a batch of records, we should produce pages from the Kafka connector by implementing ConnectorPageSource and more away from the RecordSet api.

Comment (hashhar) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaRecordSet.java:116]: PageSource is a much larger work - good point that we should consider using it since we fetch around 100k records per batch by default anyway.

In this case I'm just trying to see if we are introducing a regression or not.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaConfig.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaConfig.java:186]: It brings some inconsistency to code, but this inconsistency already exists in definitions.
In general I agree with you.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaRecordSet.java
Comment (hashhar) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaRecordSet.java:117]: Would you be open to extracting this change (just the array to list change) to a different PR and continuing there?

A simple timing test on a test which creates a sufficient number of splits (or lower number of messages per split) should be good enough to see if there's visible regression.

Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaRecordSet.java]: I've dropped that commit related to cursor refactoring. 
https://github.com/trinodb/trino/pull/14435

**Comment by ssheikin** [comment]
@hashhar I believe there no outstanding questions left here. Please approve and merge.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaRecordSet.java
Comment (hashhar) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaRecordSet.java]: with switch expression if we leave out default and are missing some enum it'll be a compile time error. Why not use that?

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java
Comment (hashhar) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java]: `TrinoException`

Or will this be wrapped in TrinoException somewhere?

File: docs/src/main/sphinx/connector/kafka.rst
Comment (hashhar) [inline docs/src/main/sphinx/connector/kafka.rst:233]:
```suggestion
configuration property and defaults to ``_``. If you have configured the property to a different value the column
names below might appear different.
```

File: plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestInternalFieldConflict.java
Comment (hashhar) [inline plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestInternalFieldConflict.java:67]: What is the top-most Exception being thrown? Is it a TrinoException?

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java
Comment (hashhar) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java:160]:
```suggestion
            throw new IllegalArgumentException("Internal Kafka column names conflict with column names from the table. "
```

**Review by hashhar — APPROVED** [review]
LGTM % comments.

**Review by Praveen2112 — APPROVED** [review]

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java:160]: `columns names` vs `column names`

I always have internal discussion how to name such variables. And I still don't know how to do that properly. Are there any rules/best practices? Why only the second word should be in multiple form?

There are many columns and each of them has only one name. 
This is not true: only one column with multiple names.

cc @mosabua

Comment (hashhar) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java:160]: english is weird - the way I remember is you can have either world plural but not both - so `column names` or `column's names` AFAIK. 🤷

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java]: No, it won't.

File: plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestInternalFieldConflict.java
Comment (ssheikin) [reply plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestInternalFieldConflict.java:67]: No, it was not.
https://github.com/trinodb/trino/pull/14224#discussion_r986478030

Comment (ssheikin) [reply plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestInternalFieldConflict.java:67]: As I've modified it to `TrinoException`, it could be changed to assertQueryFails.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaInternalFieldManager.java
Comment (wendigo) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaInternalFieldManager.java:214]: maybe instead of adding a prefix here, let's check if prefix is added on the getFieldByName() ?

**Review by wendigo — CHANGES_REQUESTED** [review]

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaInternalFieldManager.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaInternalFieldManager.java:166]: Why does it matter here?

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:99]: Same question as https://github.com/trinodb/trino/pull/14224#discussion_r987677766

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaInternalFieldManager.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaInternalFieldManager.java:214]: That would mean that `InternalField#getColumnName` returns not the actual name, but only a `root` of the name, so actually it's not usable if used alone without making prefix visible outside.

May be there is a way of general improvement of usages of `getColumnName` but I'd prefer not to.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java
Comment (hashhar) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:99]: cc: @losipiuk Do we have some convention about using interfaces (or superclasses) instead of concrete ones for local vars? I've received this comment in past from you as well - but probably we can configure Modernizer instead of manually reviewing and hoping we catch them?

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaConfig.java
Comment (hashhar) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaConfig.java:187]: To be fair any valid UTF-8 value would work without problem but some values would probably require quoting the column names.

**Review by hashhar — APPROVED** [review]

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java:99]: cc @ksobolew

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaConfig.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaConfig.java:187]: I think it's good enough for now. In any case these values are Runtime Only and not stored anywhere.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java
Comment (mosabua) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java]: single in both is correct..

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaConfig.java
Comment (hashhar) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaConfig.java:187]: They can get stored if people do CTAS statements to some other catalog but if they set an emoji as the prefix I assume they know what they are doing. This is fine IMO.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java
Comment (ssheikin) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaFilterManager.java]: No sense to discuss here.

**Comment by ssheikin** [comment]
@wendigo please approive

File: docs/src/main/sphinx/connector/kafka.rst
Comment (mosabua) [inline docs/src/main/sphinx/connector/kafka.rst:232]: Change this to a normal sentence

The internal column prefix is

File: docs/src/main/sphinx/connector/kafka.rst
Comment (mosabua) [inline docs/src/main/sphinx/connector/kafka.rst:234]: A different prefix affects the internal column names in the following sections. For example, a value of `internal_` changes the partition ID column name from `internal_partition_id` to `internal_partition_id`.

**Review by wendigo — APPROVED** [review]

**Review by hashhar — APPROVED** [review]
Thanks, will merge once CI is done.

**Comment by ssheikin** [comment]
image
Restarting checks

**Pull request merged**
